src/esm.ts

```diff
@@ -175,26 +176,33 @@ export function createEsmHooks(tsNodeService: Service) {
       }
     }
 
-    let candidateSpecifier: string | undefined;
-    while ((candidateSpecifier = candidateSpecifiers.shift())) {
+    for (let i = 0; i < candidateSpecifiers.length; i++) {
       try {
         return await nodeResolveImplementation.defaultResolve(
-          candidateSpecifier,
+          candidateSpecifiers[i],
           context,
           defaultResolve
         );
       } catch (err) {
-        const isNotFoundError = (<any>err).code === 'ERR_MODULE_NOT_FOUND';
-        if (isNotFoundError && candidateSpecifiers.length > 0) {
-          continue;
-        } else {
+        const isNotFoundError = (err as any).code === 'ERR_MODULE_NOT_FOUND';
+        if (!isNotFoundError) {
           throw err;
+        } else if (i == candidateSpecifiers.length - 1) {
+          throw new MappedModuleNotFound(
+            specifier,
+            context.parentURL,
+            candidateSpecifiers
+          );
```

Comment:
This looks great.  Should we special-case it to throw `err` when `candidateSpecifiers.length === 1`?